### PR TITLE
[WIP] Trade Widget

### PR DESCRIPTION
**Pull request by dekz**
## Trade Widget

trade-widget

### TODO
 - [ ] SRA Liquidity Provider
 - [ ] Contracts v2
 - [ ] Fee abstraction
 - [ ] NFTs
 - [ ] Sell

### CODE REVIEW

File: packages/trade-widget/src/App.css [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+.App-intro {
+  font-size: large;
+}
+
+@keyframes App-logo-spin {
+  from { transform: rotate(0deg); }
+  to { transform: rotate(360deg); }
+}
```
Comment (fabioberger) [inline packages/trade-widget/src/App.css]: I'm really not a fan of CSS. It's too easy to shoot yourself in the foot and have designs that are brittle because changing the CSS styling of a class changes the styling in many disparate places in the app. Can we continue the convention of in-lining the styling within the React components where used?

File: packages/trade-widget/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    Column,
+    Columns,
+    Container,
+    Content,
+    Control,
+    Field,
+    Input,
+    Label,
+    TextArea,
+} from 'bloomer';
```
Comment (fabioberger) [inline packages/trade-widget/src/App.tsx]: This seems like a combination of [BassCss](http://basscss.com/) and [Material-ui](http://www.material-ui.com/), two libraries we've already adopted heavily in website and react-docs, react-shared. Can you please articulate what makes `bloomer` a 10x improvement over these? If it isn't, I'd greatly prefer we continue to keep our tooling uniform.

File: packages/trade-widget/src/App.tsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import registerServiceWorker from './registerServiceWorker';
+
+import { FixedProvider } from './liquidity_providers/fixed_provider';
+import { Dispatcher } from './redux/dispatcher';
+import { reducer, State } from './redux/reducer';
+import { BlockchainSaga } from './sagas/blockchain_saga';
+import { AssetToken } from './types';
+
+const TEST_NETWORK_ID = 50;
+const TEST_RPC = 'http://127.0.0.1:8545';
```
Comment (fabioberger) [inline packages/trade-widget/src/App.tsx]: This should be in a `constants` file.

File: packages/trade-widget/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+const providerEngine = new Web3ProviderEngine();
+providerEngine.addProvider(new InjectedWeb3Subprovider((window as any).web3.currentProvider));
```
Comment (fabioberger) [inline packages/trade-widget/src/App.tsx]: What if there is no injected web3 instance on the `window` object?

File: packages/trade-widget/src/App.tsx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+providerEngine.addProvider(new RedundantRPCSubprovider([TEST_RPC]));
+providerEngine.start();
+
+const web3Wrapper = new Web3Wrapper(providerEngine);
+const zeroEx = new ZeroEx(providerEngine, { networkId: TEST_NETWORK_ID });
+
+const composeEnhancers = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
+const store: ReduxStore<State> = createStore(reducer, composeEnhancers(applyMiddleware(thunk)));
```
Comment (fabioberger) [inline packages/trade-widget/src/App.tsx]: Really not a big fan of Thunks. Even Gaeron (creator) said it should only be used for small test projects... It introduces callback hell and doesn't handle well complex async requests.

File: packages/trade-widget/src/App.tsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    render() {
+        return (
+            <Provider store={store}>
+                <Container style={{ marginTop: 20, marginLeft: 20 }}>
+                    <Content>
+                        <Columns>
+                            <Column isSize={{ mobile: 12, default: '1/4' }}>
+                                <Card>
+                                    <CardHeaderTitle>
+                                        <Label isSize={'small'}>0x TRADE WIDGET</Label>
```
Comment (fabioberger) [inline packages/trade-widget/src/App.tsx]: `isSize="small"`

File: packages/trade-widget/src/artifacts/artifacts/Forwarder.json [hunk 7]
```diff
@@ -0,0 +1,71 @@
+{
```
Comment (fabioberger) [inline packages/trade-widget/src/artifacts/artifacts/Forwarder.json]: Why do we have the Forwarder artifacts twice?

File: packages/trade-widget/src/components/AccountBlockie/AccountBlockie.tsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+import { AssetToken } from '../../types';
+
+import { Blockie } from './Blockie';
+
+const ETH_DECIMAL_PLACES = 18;
+
+/**
+ * AccountBlockie properties.
+ */
+interface AccountBlockieProps {
```
Comment (fabioberger) [inline packages/trade-widget/src/components/AccountBlockie/AccountBlockie.tsx]: Can we come up with a better name? `Blockie` doesn't mean anything to most people. Why not: `Identicon`?

File: packages/trade-widget/src/components/AccountBlockie/AccountBlockie.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            ? ZeroEx.toUnitAmount(weiBalance, ETH_DECIMAL_PLACES)
+                  .toFixed(4)
+                  .toString()
+            : '';
+    }
+    private _renderTokenBalance(): string {
+        const { tokenBalance } = this.props;
+        return tokenBalance
+            ? ZeroEx.toUnitAmount(tokenBalance, ETH_DECIMAL_PLACES)
+                  .toFixed(0)
```
Comment (fabioberger) [inline packages/trade-widget/src/components/AccountBlockie/AccountBlockie.tsx]: Why are tokens rendered without decimals? This won't suit all tokens...

File: packages/trade-widget/src/components/AccountBlockie/AccountBlockie.tsx [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                  .toString()
+            : '';
+    }
+    private _renderMinimisedAccount(): string {
+        //   0xea95a7...609353b2
+        const account = this.props.account;
+        const initial = account.substring(0, 6);
+        const end = account.substring(account.length - 6, account.length);
+        const minimised = `${initial}...${end}`;
+        return minimised;
```
Comment (fabioberger) [inline packages/trade-widget/src/components/AccountBlockie/AccountBlockie.tsx]: This is already implemented in `website`. See: https://github.com/0xProject/0x-monorepo/blob/development/packages/website/ts/utils/utils.ts#L176

File: packages/trade-widget/src/components/AccountBlockie/Blockie.tsx [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+}
+
+/* tslint:disable:member-access prefer-function-over-method */
+class Blockie extends React.Component<IdenticonPropTypes> {
+    public identicon: any;
+    constructor(props: IdenticonPropTypes) {
+        super(props);
+        this.generateIdenticon = this.generateIdenticon.bind(this);
+    }
+
```
Comment (fabioberger) [inline packages/trade-widget/src/components/AccountBlockie/Blockie.tsx]: Remove spaces between methods in class.

File: packages/trade-widget/src/components/AccountBlockie/Blockie.tsx [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    scale?: number;
+    color?: string;
+    bgColor?: string;
+    spotColor?: string;
+    [key: string]: any;
+}
+
+/* tslint:disable:member-access prefer-function-over-method */
+class Blockie extends React.Component<IdenticonPropTypes> {
+    public identicon: any;
```
Comment (fabioberger) [inline packages/trade-widget/src/components/AccountBlockie/Blockie.tsx]: any?

File: packages/trade-widget/src/components/AccountBlockie/Blockie.tsx [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    seed: string;
+    size?: number;
+    scale?: number;
+    color?: string;
+    bgColor?: string;
+    spotColor?: string;
+    [key: string]: any;
+}
+
+/* tslint:disable:member-access prefer-function-over-method */
```
Comment (fabioberger) [inline packages/trade-widget/src/components/AccountBlockie/Blockie.tsx]: Let's try and avoid disabling TSLint rules for an entire class.

File: packages/trade-widget/src/components/AccountBlockie/Blockie.tsx [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+        const aProps = Object.getOwnPropertyNames(prevProps);
+        const bProps = Object.getOwnPropertyNames(nextProps);
+
+        if (aProps.length !== bProps.length) {
+            return false;
+        }
+
+        // tslint:disable-next-line:prefer-for-of
+        for (let i = 0; i < aProps.length; i++) {
+            const propName = aProps[i];
```
Comment (fabioberger) [inline packages/trade-widget/src/components/AccountBlockie/Blockie.tsx]: Why not use `for of` here?

File: packages/trade-widget/src/components/AccountBlockie/Blockie.tsx [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    generateIdenticon(options: any) {
+        // NOTE --  Majority of this code is referenced from: https://github.com/alexvandesande/blockies
+        //          Mostly to ensure congruence to Ethereum Mist's Identicons
```
Comment (fabioberger) [inline packages/trade-widget/src/components/AccountBlockie/Blockie.tsx]: Why not just use `https://github.com/0xProject/0x-monorepo/blob/development/packages/website/ts/components/ui/identicon.tsx`?

File: packages/trade-widget/src/components/AccountBlockie/Blockie.tsx [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+        const scale = opts.scale || 4;
+        const seed = opts.seed || Math.floor(Math.random() * Math.pow(10, 16)).toString(16);
+
+        seedrand(seed);
+
+        const color = opts.color || createColor();
+        const bgcolor = opts.bgColor || createColor();
+        const spotcolor = opts.spotColor || createColor();
+        const imageData = createImageData(size);
+        const canvas = setCanvas(this.identicon, imageData, color, scale, bgcolor, spotcolor);
```
Comment (fabioberger) [inline packages/trade-widget/src/components/AccountBlockie/Blockie.tsx]: This copy-pasted code needs to go.

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+    TokenPair,
+} from '../../types';
+import AccountBlockie from '../AccountBlockie';
+import TokenSelector from '../TokenSelector';
+
+/* tslint:disable:prefer-function-over-method member-access */
+
+interface BuyWidgetProps {
+    onTokenChange?: (token: string) => any;
+    onAmountChange?: (amount: BigNumber) => any;
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: In what scenario would these props not be defined?

Comment (fabioberger) [reply packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: Also avoid using `any` please.

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+    zeroEx: ZeroEx;
+    dispatcher: Dispatcher;
+    quote: Quote;
+}
+
+interface BuyWidgetState {
+    amount?: BigNumber;
+}
+
+const ETH_DECIMAL_PLACES = 18;
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: Should be on a `constants` module.

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+class BuyWidget extends React.Component<BuyWidgetProps, BuyWidgetState> {
+    private _forwarder: ForwarderWrapper;
+    constructor(props: any) {
+        super(props);
+        this.state = {
+            amount: new BigNumber(1),
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: Do we want this to default to `1`? Also, rename `amount`. Amount is base units? units? Unclear.

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+        };
+
+        this._handleAmountChange = this._handleAmountChange.bind(this);
+        this._handleSubmitAsync = this._handleSubmitAsync.bind(this);
+    }
+
+    public async componentDidUpdate() {
+        await this._quoteSelectedTokenAsync();
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: On every update this is called? That could end up being a **ton** of times. Can you explain the reasoning?

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    render() {
+        const { address, weiBalances, tokenBalances, selectedToken, quote } = this.props;
+        const tokenBalance = tokenBalances[address] ? tokenBalances[address][selectedToken] : new BigNumber(0);
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: `_.isUndefined(tokenBalances[address])`

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+        const weiBalance = weiBalances[address] || new BigNumber(0);
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: ditto above.

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+                        isLoading={!this._hasQuoteForSelectedToken()}
+                        isFullWidth={true}
+                        isColor="info"
+                        onClick={this._handleSubmitAsync}
+                    >
+                        BUY TOKENS
+                    </Button>
+                </Field>
+                <Field style={{ marginTop: 20 }} isGrouped={'centered'}>
+                    <img style={{ marginLeft: '0px', height: '20px' }} src="/images/powered.png" />
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: For in-line, if it's `px` you want, you can simply omit the suffix: `height: 20, `

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+    private _handleAmountChange(event: any) {
+        event.preventDefault();
+        const rawValue = event.target.value;
+        let value: undefined | BigNumber;
+        if (!_.isUndefined(rawValue) && !_.isEmpty(rawValue)) {
+            const ethValue = new BigNumber(rawValue);
+            const fillAmount = ZeroEx.toBaseUnitAmount(ethValue, ETH_DECIMAL_PLACES);
+            value = fillAmount;
+            this.setState((prev, props) => {
+                return { ...prev, amount: value };
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: `baseUnitAmount`

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+    private _handleTokenSelected(token: AssetToken) {
+        this.props.dispatcher.updateSelectedToken(token);
+    }
+
+    private async _fillOrderAsync(
+        takerAddress: string,
+        fillAmount: BigNumber,
+        signedOrder: SignedOrder,
+    ): Promise<string> {
+        const txHash = await this._getForwarder().fillOrderAsync(signedOrder, fillAmount, takerAddress);
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: What happens if the user rejects the tx signing? Network not available? Need error handling.

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+        this.props.dispatcher.transactionMined(receipt);
+        return txHash;
+    }
+
+    private _getForwarder(): ForwarderWrapper {
+        const artifactJSONs = _.values(artifacts);
+        const abiArrays = _.map(artifactJSONs, artifact => artifact.networks[this.props.networkId].abi);
+        const abiDecoder = new AbiDecoder(abiArrays);
+        this._forwarder = new ForwarderWrapper(this.props.web3Wrapper, this.props.networkId, abiDecoder);
+        return this._forwarder;
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: This stuff should only happen once right?

File: packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx [hunk 27]
```diff
@@ -0,0 +1,207 @@
...
+        const { selectedToken, quote } = this.props;
+        if (!this._hasQuoteForSelectedToken()) {
+            const tokenPair: TokenPair = { maker: selectedToken, taker: AssetToken.WETH };
+            this.props.dispatcher.quoteRequested(amount, tokenPair);
+        }
+    }
+
+    private _hasQuoteForSelectedToken() {
+        const { quote, selectedToken } = this.props;
+        if (_.isUndefined(quote) || quote.pair.maker !== selectedToken) {
```
Comment (fabioberger) [inline packages/trade-widget/src/components/BuyWidget/BuyWidget.tsx]: Not checking that `quote.pair` exists.

File: packages/trade-widget/src/components/TokenSelector/TokenSelector.tsx [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+import * as ReactDOM from 'react-dom';
+
+import { AssetToken } from '../../types';
+
+import TokenSelectorItem from '../TokenSelectorItem';
+
+import { definedTokensToIcon } from '../TokenSelectorItem/TokenSelectorItem';
+
+interface TokenSelectorPropTypes {
+    onChange?: (token: AssetToken) => void;
```
Comment (fabioberger) [inline packages/trade-widget/src/components/TokenSelector/TokenSelector.tsx]: Is there somewhere this component is used an `onChange` is not specified? If not, make it required.

File: packages/trade-widget/src/components/TokenSelector/TokenSelector.tsx [hunk 29]
```diff
@@ -0,0 +1,107 @@
...
+    selectedToken: AssetToken;
+}
+
+interface TokenMetadata {
+    id: string;
+    symbol: AssetToken;
+    description: string;
+}
```
Comment (fabioberger) [inline packages/trade-widget/src/components/TokenSelector/TokenSelector.tsx]: Move to types file.

File: packages/trade-widget/src/components/TokenSelector/TokenSelector.tsx [hunk 30]
```diff
@@ -0,0 +1,107 @@
...
+class TokenSelector extends React.Component<TokenSelectorPropTypes> {
+    // tslint:disable-next-line:underscore-private-and-protected
+    private static defaultProps = {
+        // tslint:disable-next-line:no-empty
+        onChange: () => {},
```
Comment (fabioberger) [inline packages/trade-widget/src/components/TokenSelector/TokenSelector.tsx]: `_.noop`

File: packages/trade-widget/src/components/TokenSelector/TokenSelector.tsx [hunk 31]
```diff
@@ -0,0 +1,107 @@
...
+    };
+    private _tokens: TokenMetadata[];
+    constructor(props: TokenSelectorPropTypes) {
+        super(props);
+        this.handleItemSelected = this.handleItemSelected.bind(this);
+        this._tokens = [
+            { symbol: AssetToken.ZRX, id: AssetToken.ZRX, description: '0x Token' },
+            { symbol: AssetToken.BAT, id: AssetToken.BAT, description: 'Basic Attention Token' },
```
Comment (fabioberger) [inline packages/trade-widget/src/components/TokenSelector/TokenSelector.tsx]: Symbol and id seem duplicitous. What's the difference? Also, should be by relying on symbols as the id for tokens? They are not guaranteed to be unique. Really we should be using their contract addresses.

File: packages/trade-widget/src/components/TokenSelectorItem/TokenSelectorItem.tsx [hunk 32]
```diff
@@ -0,0 +1,76 @@
...
+    ),
+    ZRX: <span className={'icon-ZeroEx is-left icon'} />,
+};
+
+import { AssetToken } from '../../types';
+interface TokenSelectorItemPropTypes {
+    id: string;
+    description: string;
+    symbol: AssetToken;
+    isActive?: boolean;
```
Comment (fabioberger) [inline packages/trade-widget/src/components/TokenSelectorItem/TokenSelectorItem.tsx]: Why optional?

File: packages/trade-widget/src/components/TokenSelectorItem/TokenSelectorItem.tsx [hunk 33]
```diff
@@ -0,0 +1,76 @@
...
+            <span className="path3" />
+            <span className="path4" />
+            <span className="path5" />
+        </span>
+    ),
+    ZRX: <span className={'icon-ZeroEx is-left icon'} />,
+};
+
+import { AssetToken } from '../../types';
+interface TokenSelectorItemPropTypes {
```
Comment (fabioberger) [inline packages/trade-widget/src/components/TokenSelectorItem/TokenSelectorItem.tsx]: Please add `.prettierrc`... Or why isn't it picking up stylistic issues?

File: packages/trade-widget/src/components/TokenSelectorItem/TokenSelectorItem.tsx [hunk 34]
```diff
@@ -0,0 +1,76 @@
...
+    description: string;
+    symbol: AssetToken;
+    isActive?: boolean;
+    onClick: (item: TokenSelectorItem) => void;
+}
+
+class TokenSelectorItem extends React.Component<TokenSelectorItemPropTypes> {
+    constructor(props: any) {
+        super(props);
+    }
```
Comment (fabioberger) [inline packages/trade-widget/src/components/TokenSelectorItem/TokenSelectorItem.tsx]: Remove constructor since not used.

File: packages/trade-widget/src/components/TokenSelectorItem/TokenSelectorItem.tsx [hunk 35]
```diff
@@ -0,0 +1,76 @@
...
+
+    // tslint:disable-next-line:member-access
+    public handleClicked(e: any) {
+        const { onClick } = this.props;
+        onClick(this);
+    }
```
Comment (fabioberger) [inline packages/trade-widget/src/components/TokenSelectorItem/TokenSelectorItem.tsx]: Remove this method. Can call `props.onClick` directly. Also, passing `this` as a prop is a big red flag. Instead just pass the symbol... or rather the token contract address (unique).

File: packages/trade-widget/src/contract_wrappers/generated/base_contract.ts [hunk 36]
```diff
@@ -0,0 +1,33 @@
...
+            const estimatedGas = await estimateGasAsync(txData);
+            txDataWithDefaults.gas = estimatedGas;
+        }
+        return txDataWithDefaults;
+    }
+    constructor(web3ContractInstance: Web3.ContractInstance, defaults: Partial<TxData>) {
+        this._web3ContractInstance = web3ContractInstance;
+        this._defaults = defaults;
+    }
+}
```
Comment (fabioberger) [inline packages/trade-widget/src/contract_wrappers/generated/base_contract.ts]: This should all not be copy-pasted.

File: packages/trade-widget/src/liquidity_providers/fixed_provider.ts [hunk 37]
```diff
@@ -0,0 +1,72 @@
+import { SignedOrder } from '0x.js';
+import { BigNumber } from '@0xproject/utils';
+// tslint:disable-next-line:no-var-requires
+const delay = require('delay');
```
Comment (fabioberger) [inline packages/trade-widget/src/liquidity_providers/fixed_provider.ts]: Can you not import this another way without breaking the Tslint rule?

File: packages/trade-widget/src/logo.svg [hunk 38]
```diff
@@ -0,0 +1,7 @@
...
+        <circle cx="420.9" cy="296.5" r="45.7"/>
+        <path d="M520.5 78.1z"/>
+    </g>
+</svg>
```
Comment (fabioberger) [inline packages/trade-widget/src/logo.svg]: All SVG's should be in a `public/svgs` directory. All assets should be in a `public` dir.

File: packages/trade-widget/src/redux/dispatcher.ts [hunk 39]
```diff
@@ -0,0 +1,85 @@
...
+                dispatch({
+                    data: { quote },
+                    type: ActionTypes.QuoteReceived,
+                });
+            } catch (err) {
+                dispatch({
+                    data: { amount, pair },
+                    type: ActionTypes.QuoteRequestFailed,
+                });
+            }
```
Comment (fabioberger) [inline packages/trade-widget/src/redux/dispatcher.ts]: Let's please not use thunks.

File: packages/trade-widget/src/sagas/blockchain_saga.ts [hunk 40]
```diff
@@ -0,0 +1,99 @@
...
+        this._startPolling().catch(console.log);
+        store.subscribe(this._handleStateChange.bind(this));
+        this._handleStateChange();
+    }
+    // The intention here is to stop any long running processes
+    // if they are not relevant anymore. I.e watching for balance changes
+    // on an old network.
+    private _handleStateChange() {
+        const nextState = this._store.getState();
+        const lastAction = nextState.lastAction;
```
Comment (fabioberger) [inline packages/trade-widget/src/sagas/blockchain_saga.ts]: Unused.

File: packages/trade-widget/src/sagas/blockchain_saga.ts [hunk 41]
```diff
@@ -0,0 +1,99 @@
...
+        if (nextState !== this._prevState) {
```
Comment (fabioberger) [inline packages/trade-widget/src/sagas/blockchain_saga.ts]: Does this comparison work for instances of a class?

File: packages/trade-widget/src/sagas/blockchain_saga.ts [hunk 42]
```diff
@@ -0,0 +1,99 @@
...
+        this._web3Wrapper = web3Wrapper;
+        this._store = store;
+        this._zeroEx = zeroEx;
+        this._startPolling().catch(console.log);
+        store.subscribe(this._handleStateChange.bind(this));
+        this._handleStateChange();
+    }
+    // The intention here is to stop any long running processes
+    // if they are not relevant anymore. I.e watching for balance changes
+    // on an old network.
```
Comment (fabioberger) [inline packages/trade-widget/src/sagas/blockchain_saga.ts]: Comment doesn't seem related to this method.

File: packages/trade-widget/src/sagas/blockchain_saga.ts [hunk 43]
```diff
@@ -0,0 +1,99 @@
...
+
+    private async _checkNetworkIdAsync() {
+        const currentNetworkId = await this._web3Wrapper.getNetworkIdAsync();
+        const networkId = this._store.getState().networkId;
+        if (networkId !== currentNetworkId) {
+            this._dispatch.updateNetworkId(currentNetworkId);
+        }
+    }
+
+    private async _checkUserAccountAsync() {
```
Comment (fabioberger) [inline packages/trade-widget/src/sagas/blockchain_saga.ts]: I'd prefer naming this `_updateUserAccountIfChanged`

File: packages/trade-widget/src/sagas/blockchain_saga.ts [hunk 44]
```diff
@@ -0,0 +1,99 @@
...
+        const userAccounts = await this._web3Wrapper.getAvailableAddressesAsync();
+        const currentUserAccount = userAccounts[0];
+        const userAccount = this._store.getState().userAddress;
+        if (userAccount !== currentUserAccount) {
+            this._dispatch.updateUserAddress(currentUserAccount);
+        }
+    }
+}
```
Comment (fabioberger) [inline packages/trade-widget/src/sagas/blockchain_saga.ts]: This is already an improvement over the `BlockchainWatcher` we have in `website`. What I like is that we can access the store, rather than duplicating those values as instance variables.

File: packages/trade-widget/src/sagas/blockchain_saga.ts [hunk 45]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    private async _checkTokenBalanceAsync() {
+        const currentToken = this._store.getState().selectedToken;
+        const currentUser = this._store.getState().userAddress;
+        const userTokenBalances = this._store.getState().userTokenBalances[currentUser];
+        let tokenBalance = new BigNumber(0);
+
+        // TODO this won't work for the many tokens and NFTs not in the registry
+        const tokenAddress = await this._zeroEx.tokenRegistry.getTokenAddressBySymbolIfExistsAsync(currentToken);
```
Comment (fabioberger) [inline packages/trade-widget/src/sagas/blockchain_saga.ts]: This is also super inefficient. This should be fetched once and cached. We don't expect to be updating the tokenRegistry frequently and the client shouldn't expect real-time updates... In the off chance this does update, they should just re-fresh the page.

Comment (fabioberger) [reply packages/trade-widget/src/sagas/blockchain_saga.ts]: Thinking more long term though, really the list of desired tokens tradeable on the Trade Widget should be passed in at component instantiation. Different sites will want to cater to different audiences...

File: packages/trade-widget/src/sagas/blockchain_saga.ts (hunk 44 above)
Comment (fabioberger) [reply packages/trade-widget/src/sagas/blockchain_saga.ts]: What this component still doesn't do is deal with dependent updates. If the networkId changed, we need to re-fetch the tokens, balances, allowances, **before** dispatching an action. Otherwise the store will be in an inconsistent state.

File: packages/trade-widget/src/types.ts [hunk 46]
```diff
@@ -0,0 +1,65 @@
+import { Order, SignedOrder } from '0x.js';
+import { BigNumber } from '@0xproject/utils';
+import * as Web3 from 'web3';
+
+export type ArtifactContractName = 'Forwarder';
```
Comment (fabioberger) [inline packages/trade-widget/src/types.ts]: This should be in a `constants.ts` or `configs.ts` file as `ARTIFACT_CONTRACT_NAME`.

File: packages/trade-widget/src/types.ts [hunk 47]
```diff
@@ -0,0 +1,65 @@
...
+            address: string;
+            abi: Web3.ContractAbi;
+        };
+    };
+}
+
+export enum AssetToken {
+    ZRX = 'ZRX',
+    BAT = 'BAT',
+    WETH = 'WETH',
```
Comment (fabioberger) [inline packages/trade-widget/src/types.ts]: Enum convention is capitalized names, all caps values. E.g

```
export enum AssetToken {
  Zrx = 'ZRX',
  Bat = 'BAT',
  ...
}
```

File: packages/trade-widget/src/zeroex_artifacts.ts [hunk 48]
```diff
@@ -0,0 +1,16 @@
+/* tslint:disable */
```
Comment (fabioberger) [inline packages/trade-widget/src/zeroex_artifacts.ts]: Let's rename this to `zero_ex_artifacts.ts`

File: packages/trade-widget/truffle.js [hunk 49]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
+  contracts_build_directory: "../contracts/artifacts",
+  networks: {
+    development: {
+      host: "127.0.0.1",
+      port: 8545,
+      network_id: "*" // Match any network id
+    }
+  }
+};
```
Comment (fabioberger) [inline packages/trade-widget/truffle.js]: Eeeek... Are you using Truffle?

File: packages/trade-widget/tsconfig.json [hunk 50]
```diff
@@ -0,0 +1,26 @@
+{
+    "extends": "../../tsconfig",
+    "compilerOptions": {
+        "outDir": "lib",
+        "jsx": "react",
+        "noImplicitThis": false,
+        "baseUrl": "./",
+        "allowJs": true,
+        "strictNullChecks": false,
```
Comment (fabioberger) [inline packages/trade-widget/tsconfig.json]: Set this to `true`.

File: packages/trade-widget/truffle.js (hunk 49 above)
Comment (dekz) [reply packages/trade-widget/truffle.js]: Oh no this is from experimenting with trying to get the contract working with truffles new debugger, this is an unused file :P
